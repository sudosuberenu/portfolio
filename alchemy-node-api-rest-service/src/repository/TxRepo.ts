import { AssetTransfersResult } from 'alchemy-sdk';

import { TxModel } from '../model/Tx';
import Logger from '../core/Logger';

export default class TxRepo {
  public static async fetchFromBlock(blockNumber: number): Promise<AssetTransfersResult[]> {
    const txs = await TxModel.fetchFromBlock(blockNumber);
    Logger.info(`${txs.length} txs has been fetched from Alchemy`);
    return txs;
  }

  public static async createBatch(txs: AssetTransfersResult[]): Promise<boolean> {
    let promises: Promise<boolean>[] = [];
    
    txs.forEach(tx => {
      promises.push(this.create(tx));
    })

    await Promise.all(promises);
    return true;
  }

  public static async create(tx: AssetTransfersResult): Promise<boolean> {
    await TxModel.create(tx);
    Logger.info(`tx ${tx.hash} stored in owned database`);
    return true;
  }

  public static async getDaiTxs(offset: number, page: string): Promise<any> {
    const response = await TxModel.getDaiTxs(offset, page);
    Logger.info(`Fetched ${response.txs.length} DAI txs from 3rd repo`);
    return response;
  }

  public static async getTxsFromSender(from: string): Promise<any> {
    const response = await TxModel.getTxsFromSender(from);
    Logger.info(`Fetched ${response.txs.length} txs from sender ${from} from 3rd repo`);
    return response;
  }

  public static async getTxsFromReceipt(to: string): Promise<any> {
    const response = await TxModel.getTxsFromReceipt(to);
    Logger.info(`Fetched ${response.txs.length} txs from receipt ${to} from 3rd repo`);
    return response;
  }
}