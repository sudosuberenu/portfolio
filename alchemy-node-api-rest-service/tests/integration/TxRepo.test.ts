import TxRepo from "../../src/repository/TxRepo";

describe('TxRepo', () => {
  describe('fetchFromBlock', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    })

    it('When given correct blockNumber then return correct txs', async () => {
      const expectedResult = {
        "asset": "DAI", "blockNum": "0xbc614e", 
        "category": "erc20", "erc1155Metadata": null, "erc721TokenId": null, 
        "from": "0x503828976d22510aad0201ac7ec88293211d23da", "hash": "0x59ab07eddb9093b18466f4fe39544fb11dc028790bc675382e121c5851092c3c", 
        "rawContract": {"address": "0x6b175474e89094c44da98b954eedeac495271d0f", "decimal": "0x12", "value": "0x0a6af0535661d3d400"}, 
        "to": "0x8213001dc18a285c8527f86cf0b4267ce6bd0616", "tokenId": null, "uniqueId": "0x59ab07eddb9093b18466f4fe39544fb11dc028790bc675382e121c5851092c3c:log:14", "value": 192.17319133}

      const result = await TxRepo.fetchFromBlock(12345678);

      expect(result.length).toBe(1000);
      expect(result[0]).toStrictEqual(expectedResult);
    })

    it('When given incorrect blockNumber then throw error', async () => {   
      await expect(TxRepo.fetchFromBlock(-1)).rejects.toThrowErrorMatchingSnapshot();
    })

  })
})
