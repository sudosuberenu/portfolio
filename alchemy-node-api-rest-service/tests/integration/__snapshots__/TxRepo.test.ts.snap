// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TxRepo fetchFromBlock When given incorrect blockNumber then throw error 1`] = `"processing response error (body=\\"{\\\\\\"jsonrpc\\\\\\":\\\\\\"2.0\\\\\\",\\\\\\"id\\\\\\":43,\\\\\\"error\\\\\\":{\\\\\\"code\\\\\\":-32602,\\\\\\"message\\\\\\":\\\\\\"Invalid string block in object param: fromBlock\\\\\\"}}\\", error={\\"code\\":-32602}, requestBody=\\"{\\\\\\"method\\\\\\":\\\\\\"alchemy_getAssetTransfers\\\\\\",\\\\\\"params\\\\\\":[{\\\\\\"fromBlock\\\\\\":\\\\\\"-0x01\\\\\\",\\\\\\"contractAddresses\\\\\\":[\\\\\\"0x6B175474E89094C44Da98b954EedeAC495271d0F\\\\\\"],\\\\\\"excludeZeroValue\\\\\\":false,\\\\\\"category\\\\\\":[\\\\\\"erc20\\\\\\"]}],\\\\\\"id\\\\\\":43,\\\\\\"jsonrpc\\\\\\":\\\\\\"2.0\\\\\\"}\\", requestMethod=\\"POST\\", url=\\"https://eth-mainnet.g.alchemy.com/v2/L9TmAZ6n5GpjN64JU_-xJbRxTumLdyCm\\", code=SERVER_ERROR, version=web/5.7.0)"`;
